### ISSUE TITLE
Failed SSL connections on MS Windows

### ISSUE THREAD

**Issue by Zapotek** [puma/puma#829]
Hello,

I'm getting failed SSL connections when running on MS Windows, while the same setup seems to be working fine on Linux.

My setup is:
* Windows 10 64bit
* `ruby 2.2.3p173 (2015-08-18 revision 51636) [x64-mingw32]`
* DevKit-mingw64-64-4.7.2-20130224-1432-sfx
* [OpenSSL 1.0.1m](http://dl.bintray.com/oneclick/OpenKnapsack/x64/:openssl-1.0.1m-x64-windows.tar.lzma)
* Puma 2.15.3

## Installation

I copied the `.dlls` under OpenSSL's `bin/` to Ruby's `bin/` (I had to replace the ones already there) before installing Puma with:

```
> gem install puma -- --with-opt-dir=C:\openssl
Fetching: puma-2.15.3.gem (100%)
Temporarily enhancing PATH to include DevKit...
Building native extensions with: '--with-opt-dir=C:\openssl'
This could take a while...
Successfully installed puma-2.15.3
Parsing documentation for puma-2.15.3
Installing ri documentation for puma-2.15.3
Done installing documentation for puma after 5 seconds
1 gem installed
```

## Server

```ruby
require 'puma'

class HelloWorld
    def call( env )
        [
            200,
            {
                "Content-Type" => "text/html"
            },
            ["Hello World!"]
        ]
    end
end

server = Puma::Server.new( HelloWorld.new )
server.min_threads = 0
server.max_threads = 16

options = {
    bind: '0.0.0.0',
    port: 9999
}

ctx = Puma::MiniSSL::Context.new
ctx.key  = "spec/support/pems/server/key.pem"
ctx.cert = "spec/support/pems/server/cert.pem"

server.binder.add_ssl_listener( options[:bind], options[:port], ctx )

puts "Listening on https://#{options[:bind]}:#{options[:port]}"

begin
    server.run.join
rescue Interrupt
    server.stop( true )
end
```

PEMs:
* [key](https://github.com/Arachni/arachni/tree/master/spec/support/pems/server/key.pem)
* [cert](https://github.com/Arachni/arachni/tree/master/spec/support/pems/server/cert.pem)

## Trigger

```
> curl -vvv https://127.0.0.1:9999/ -k
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 9999 (#0)
* TLSv1.0, TLS handshake, Client hello (1):
* TLSv1.0, TLS handshake, Server hello (2):
* TLSv1.0, TLS handshake, CERT (11):
* TLSv1.0, TLS handshake, Server key exchange (12):
* TLSv1.0, TLS handshake, Server finished (14):
* TLSv1.0, TLS handshake, Client key exchange (16):
* TLSv1.0, TLS change cipher, Client hello (1):
* TLSv1.0, TLS handshake, Finished (20):
* Unknown SSL protocol error in connection to 127.0.0.1:9999
* Closing connection 0
curl: (35) Unknown SSL protocol error in connection to 127.0.0.1:9999
```

```
curl 7.40.0 (x86_64-pc-win32) libcurl/7.40.0 OpenSSL/1.0.0o zlib/1.2.8 WinIDN libssh2/1.4.3 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap pop3 pop3s rtmp rtsp scp sftp smtp smtps telnet tftp
Features: AsynchDNS IDN Largefile SSPI Kerberos SPNEGO NTLM SSL libz
```

**Comment by Zapotek** [comment]
I also tried it with the PEMs used in [testing](https://github.com/puma/puma/blob/master/test/ssl_config.rb) and got the same results, any feedback on this would be appreciated.

If this feature is broken on Windows the [platform constraints](https://github.com/puma/puma#platform-constraints) should be updated to reflect that.

**Comment by evanphx** [comment]
Sorry for the delay in getting back to this. It should definitely work. Does it work using puma via the config file and running with `puma`?

**Comment by evanphx** [comment]
Due to the age of this issue, I'm going to go ahead and close it. If it's still an issue, please reopen. Sorry again about the reply delay.

**Comment by vais** [comment]
@evanphx this is still very much a problem on latest puma and latest Ruby that'a available on Windows (2.3.3). Puma builds on Windows without SSL by default, recompiled with SSL in the same way @Zapotek described, after which puma happily runs bound to SSL as if everything's fine, but the server does not respond when you try to hit it in the browser. Does anyone on the puma core have a Windows machine to look into this issue?

**Comment by preetpalS** [comment]
@evanphx I also experienced this issue with Puma (puma (3.9.1)) with Ruby 2.4.1 (x64) on Windows. Using Puma (puma (3.9.1-java)) with JRuby 9.1.9.0, I did not encounter this problem (the app was able to serve pages over HTTPS).

@vais If your application runs on JRuby, then you can use Puma with SSL on Windows.

The following is my Puma config:

```
bind 'ssl://192.168.1.50:443?key=cert.key&cert=cert.crt&verify_mode=none'

# Working JRuby HTTPS config
# bind 'ssl://192.168.1.50:443?keystore=cert.jks&keystore-pass=password&cert.key&cert=cert.crt&verify_mode=none'
```

Testing failing version (Puma 3.9.1 on Ruby 2.4.1 (x64)) with Curl (mingw64):

```
$ curl -vvv https://192.168.1.50/ -k
* STATE: INIT => CONNECT handle 0x600057470; line 1418 (connection #-5000)
* Added connection 0. The cache now contains 1 members
*   Trying 192.168.1.50...
* TCP_NODELAY set
* STATE: CONNECT => WAITCONNECT handle 0x600057470; line 1471 (connection #0)
* Connected to 192.168.1.50 (192.168.1.50) port 443 (#0)
* STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x600057470; line 1588 (connection #0)
* Marked for [keep alive]: HTTP default
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /usr/ssl/certs/ca-bundle.crt
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* STATE: SENDPROTOCONNECT => PROTOCONNECT handle 0x600057470; line 1602 (connection #0)
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=XX; ST=XX; L=XX; O=XX; CN=XX
*  start date: Aug 12 04:49:02 2017 GMT
*  expire date: Aug 10 04:49:02 2027 GMT
*  subject: C=XX; ST=XX; L=XX; O=XX; CN=XX
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
* STATE: PROTOCONNECT => DO handle 0x600057470; line 1623 (connection #0)
> GET / HTTP/1.1
> Host: 192.168.1.50
> User-Agent: curl/7.53.1
> Accept: */*
>
* STATE: DO => DO_DONE handle 0x600057470; line 1685 (connection #0)
* STATE: DO_DONE => WAITPERFORM handle 0x600057470; line 1812 (connection #0)
* STATE: WAITPERFORM => PERFORM handle 0x600057470; line 1822 (connection #0)
* SSL read: error:00000000:lib(0):func(0):reason(0), errno 104
* Marked for [closure]: Transfer returned error
* multi_done
* Closing connection 0
* The cache now contains 0 members
* Expire cleared
curl: (56) SSL read: error:00000000:lib(0):func(0):reason(0), errno 104
```
